daemon/inertia/auth/password.go

```diff
@@ -0,0 +1,64 @@
+package auth
+
+import (
+	"errors"
+
+	"golang.org/x/crypto/bcrypt"
+)
+
+var (
+	errSameUsernamePassword = errors.New("Username and password must be different")
+	errInvalidUsername      = errors.New("Only letters, numbers and underscores are allowed in usernames")
+	errInvalidPassword      = errors.New("Only letters, numbers and underscores are allowed in passwords, and password must be at least 5 characters")
+)
+
+func hashPassword(password string) (string, error) {
+	hash, err := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
+	if err != nil {
+		return "", errors.New("bcrypt password hashing unsuccessful: " + err.Error())
+	}
+	return string(hash), nil
+}
+
+func correctPassword(hash string, password string) bool {
+	return bcrypt.CompareHashAndPassword([]byte(hash), []byte(password)) == nil
+}
+
+// validateCredentialValues takes a username and password and verifies
+// if they are of sufficient length and if they only contain legal characters
+func validateCredentialValues(username, password string) error {
+	if username == password {
+		return errSameUsernamePassword
+	}
+	if len(password) < 5 || len(password) >= 128 {
+		return errInvalidPassword
+	}
+	if len(username) < 3 || len(username) >= 128 {
+		return errInvalidUsername
+	}
+	validChars := "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_"
+
+NEXT_USERNAME_CHAR:
+	for _, char := range username {
```

Comment:
Not sure how idiomatic this is, but I would avoid the nested loops here using a helper function:

```golang
// Returns true if `str` only contains characters [A-Z], [a-z], or '_'
func legalString(str string) bool {
	for _, c := range str {
		if (c < 'a' || c > 'z') && (c < 'A' || c > 'Z') && c != '_' {
			return false
		}
	}
	return true
}
```

Then the validity checks can look like this:

```golang
if len(password) < 5 || len(password) >= 128 || !validString(password) {
	return errInvalidPassword
}
if len(username) < 3 || len(username) >= 128 || !validString(username) {
	return errInvalidUsername
}
```

Reply:
Wow this is a very clean alternative, thanks!